vendor/github.com/containers/image/v5/docker/docker_transport.go

```diff
@@ -51,11 +51,22 @@ func ParseReference(refString string) (types.ImageReference, error) {
 	if !strings.HasPrefix(refString, "//") {
 		return nil, fmt.Errorf("docker: image reference %s does not start with //", refString)
 	}
+	// Check if ref has DigestOnlyTag suffixed to it
+	digestOnly := false
+	if strings.HasSuffix(refString, reference.DigestOnlyTag) {
+		digestOnly = true
+		refString = strings.TrimSuffix(refString, reference.DigestOnlyTag)
+	}
 	ref, err := reference.ParseNormalizedNamed(strings.TrimPrefix(refString, "//"))
 	if err != nil {
 		return nil, err
 	}
-	ref = reference.TagNameOnly(ref)
+
+	if digestOnly {
+		ref = reference.TagDigestOnly(ref)
+	} else {
+		ref = reference.TagNameOnly(ref)
+	}
```

Comment:
(Right now, I don’t see that this ever produces something different than the previous version did.)

Reply:
… oh, it does differ, actually.
- It does _something_ (adding a tag, or not changing the input at all) in the unintended matches from https://github.com/containers/podman/pull/20486#discussion_r1392992477 . …
- The way to represent “push digest-only to `example.com/repo` seems to be, right now, `//example.com/repodigest-only`. That’s rather something I wouldn’t expect.

Reply:
So this adds a `:digest-only` tag to the reference here. I kept running into a segfault when no tag was given to the reference. Usually the latest tag is given when none is specified so things move on smoothly, but for the digest-only use case we need to give it a tag that we recognize and can remove before the final push with digest only.

I wasn't seeing any other way of getting around the nil pointer error. If there is something I missed with that, please let me know.

Reply:
I don’t know which “a segfault” that is… my best guess is that `PolicyConfigurationIdentity` panics.

AFAIK `reference.Named` can represent name-only data fine; that’s what happens with `docker://example.com/repo`, before `TagNameOnly` is called, we hold a valid `IsNameOnly` reference.

Yes, the code does currently assume that `!reference.IsNameOnly(dockerReference.ref)`. If that assumption will be broken, all users of `dockerrReference` need to be adjusted to either handle it, or to refuse such values.
- `NewImageSource` / `DeleteImage` should refuse such references
-  `StringWithinTransport` and `PolicyConfiguration*` will need new support for the new values
- `tagOrDigest` should not be called with such values, probably to be ensured in the callers.

Reply:
Yes, `PolicyConfigurationIdentity` is what throws the segfault when the image is a `IsNameOnly`. I have now added a bool `isUnknownDigest` field to `dockerReference` that is set when `newReference` is called based on the results of `ParseReference`. 
The `ParseReference` function still calls `TagNameOnly` and adds the latest tag to it so we don't run into any errors with the restriction around `IsNameOnly`.
Once we get to `PutManifest`, I check for the `d.ref.isUnknownDigest` value and based on that decide whether to push with digest only or continue with the tag.

The code continues to assume `!reference.IsNameOnly(dockerReference.ref)`, so I don't think any of the updates you mentioned above are required. But please let me know if I am missing something.

Reply:
> The code continues to assume `!reference.IsNameOnly(dockerReference.ref)`, so I don't think any of the updates you mentioned above are required

But that’s not really the semantics of the new `isUnknownDigest` values. Those _are_ `IsNameOnly` (= repo specified, no tag / no digest). So all of that code needs to be updated to either handle or reject such values.

AFAICS the code currently treats `example.com/repo@@unknown-digest@@` the same as `example.com/repo:latest`, in most places, and that’s just unexpected. (The extreme case is `StringWithinTransport`; the API documents that `ParseReference(x.StringWithinTransport()) ~= x`).

Reply:
… IOW whether we store `dockerReference.ref` as a `IsNameOnly` value with no tag/digest, or as a `!IsNameOnly` value with some tag/digest which is ignored, that individual bit does not matter so much, we can choose for implementation convenience.

But the _whole_ of `dockerReference` now has a new possible kind of semantics, so everything which previously assumed that `dockerReferece` has exactly one of (tag, digest) needs to revise that assumption.

Reply:
Okay, makes sense. I have updated the dokerReference functions, but have a few questions:

1. Should `NewImage` and `NewImageDestination` also reject name only refs with unknownDigst=true? I don't think `NewImageDestination` should reject it as we want to push an image with an unknownDigest, please let me know if I am wrong.
2. And how should `StringWithinTransport` handle this? I am not sure how calling a `reference.FamiliarString` on a name only reference would be different

I have made changes based on the reviews below as well, PTAL.

Reply:
Looks good overall. Eventually unit tests for the reference changes will be necessary.

> 1. Should `NewImage` and `NewImageDestination` also reject name only refs with unknownDigst=true? I don't think `NewImageDestination` should reject it as we want to push an image with an unknownDigest, please let me know if I am wrong.

You’re right, `NewImageDestination` must not reject it.

`NewImage` calls `newImageSource`; so if the check is moved from `NewImageSource` to `newImageSource` (as suggested elsewhere for mostly aesthetic reasons), then `NewImage` doesn’t need an explicit condition.

> 2. And how should `StringWithinTransport` handle this? I am not sure how calling a `reference.FamiliarString` on a name only reference would be different

AFAICS `reference.FamiliarString(nameOnly)` does the right thing (e.g. `docker.io/library/busybox` → `"busybox"`), so `StringWithinTransport would only need to append `UnknownDigestSuffix`.